### PR TITLE
Feature/dex 577 rework theme

### PR DESCRIPTION
**Pull request by ceriksson**
DEX-577

# What this PR does

Add completely revamped theme with backwards compatibility to deprecated methods.

- [ ] fixing some UI issues
- [ ] looking to add tests where relevant

# Notable decisions & other stuff

WIP

# Instructions on how to test this

Should be available in appetize and emulator. Main thing to look for is that the UI has not been adversely affected by this update.

# Before merging

_QA_

- [ ] I manually tested this with a demo application (simulator)
- [ ] I manually tested this with a demo application (real device)
- [ ] I wrote unit tests for each new util-like function
- [ ] I wrote at least basic e2e tests for the new functionalities or new paths
- [ ] If this PR introduces UI changes, I manually tested this PR on iPhone 8 (iOS 10, 12, 14), iPhone 12, and iPhone 12 Max
- [ ] If this PR modifies the behavior of an API, I did my best to make my new API backward compatible
- [ ] If this PR modifies the API, I attempted to implement the feature with an example

_Documentation_

- [ ] If this PR modifies the API, I updated the API Reference.
- [ ] If this PR adds new options to the API, I wrote some documentation or guide in the online documentation
- [ ] If the PR modifies the API, I wrote some documentation to deprecate it

# After merging

- Make sure a new release has been pushed to Cocoapods trunk

### CODE REVIEW

File: Sources/PrimerSDK/Classes/Constants/Colors.swift [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        static let PaymentMethodSelected = UIColor.primer(lightMode: blue, darkMode: darkModeBlue)
+        static let PaymentMethodBorderDefault = UIColor.primer(lightMode: gray, darkMode: gray)
+        static let PaymentMethodBorderDisabled = UIColor.primer(lightMode: lightGray, darkMode: darkModeLightGray)
+        static let PaymentMethodBorderSelected = UIColor.primer(lightMode: blue, darkMode: darkModeBlue)
+        static let PaymentMethodTextEnabled = UIColor.primer(lightMode: black, darkMode: darkModeBlack)
+        static let PaymentMethodTextDisabled = UIColor.primer(lightMode: gray, darkMode: gray)
+        
+        // Klarna
+        static let KlarnaDefault = UIColor(red: 1, green: 0.702, blue: 0.78, alpha: 1)
+        static let KlarnaTextDefault = UIColor.primer(lightMode: black, darkMode: darkModeBlack)
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/Constants/Colors.swift:59]: [Feedback]
Does that need to be shared? It seems too specific to be in the constants

File: Sources/PrimerSDK/Classes/Data Models/Theme/Internal/PrimerTheme+Colors.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
+public extension UIColor {
+
+    static func primer(lightMode: UIColor, darkMode: UIColor) -> UIColor {
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/Data Models/Theme/Internal/PrimerTheme+Colors.swift:3]: [Feedback]
I don't believe this is a good name for a function that returns one of the two provided colors.

File: Sources/PrimerSDK/Classes/Data Models/Theme/Public/PublicThemeData.swift [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        self.subtitle = subtitle
+        self.amountLabel = amountLabel
+        self.system = system
+        self.error = error
+    }
+}
+
+public class TextThemeData {
+    public var defaultColor: UIColor?
+    public var fontsize: Int?
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/Data Models/Theme/Public/PublicThemeData.swift:104]: [Feedback]
This should be called `fontSize`. Could you update your public interface?

File: Sources/PrimerSDK/Classes/Data Models/Theme/Internal/PrimerTheme+Colors.swift (hunk 2 above)
Comment (ceriksson) [reply Sources/PrimerSDK/Classes/Data Models/Theme/Internal/PrimerTheme+Colors.swift:3]: Yeah, agree. Debating whether we should actually have it. Technically, if the merchant wants to pass in a UIColor that dynamically switches between dark mode and light mode, they can do so on their end. Wdyt?

File: Sources/PrimerSDK/Classes/Constants/Colors.swift (hunk 1 above)
Comment (ceriksson) [reply Sources/PrimerSDK/Classes/Constants/Colors.swift:59]: Was probably veering a little close to the Android way of doing things 😅. I'll see if I can reduce the footprint and just reference the raw color constants in the theme.

File: Sources/PrimerSDK/Classes/Constants/Dimensions.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    static let amountLabel = 24
+    static let body = 14
+    static let system = 12
+    static let error = 10
+    static let buttonLabel = 14
+}
+
+internal struct Layout {
+    
+    struct View {
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/Constants/Dimensions.swift:14]: refactor to class

File: Sources/PrimerSDK/Classes/Constants/Dimensions.swift [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        static let cornerRadius: CGFloat = 0.0
+        static let safeArea: CGFloat = 10.0
+    }
+
+    struct PaymentMethodButton {
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/Constants/Dimensions.swift:19]: change to just button

File: Sources/PrimerSDK/Classes/Data Models/PCI/FormType.swift [hunk 6]
```diff
@@ -46,7 +46,7 @@ enum FormType {
                 [.country(mandate.address?.countryCode)]
             ]
         case .cardForm(let theme):
-            switch theme.textFieldTheme {
+            switch theme.input.inputType {
             case .doublelined:
```
Comment (ceriksson) [inline Sources/PrimerSDK/Classes/Data Models/PCI/FormType.swift:50]: try to remove

File: Sources/PrimerSDK/Classes/User Interface/Root/PrimerNavigationBar.swift [hunk 7]
```diff
@@ -25,7 +25,7 @@ class PrimerNavigationBar: PrimerView {
     
     var rightBarButton: UIButton? {
         didSet {
-            rightBarButton?.tintColor = theme.text.system.color
+            rightBarButton?.tintColor = .systemRed
```
Comment (EvansPie) [inline Sources/PrimerSDK/Classes/User Interface/Root/PrimerNavigationBar.swift:28]: why not our red error color?

**Review by EvansPie — APPROVED** [review]

File: Sources/PrimerSDK/Classes/User Interface/Root/PrimerNavigationBar.swift (hunk 7 above)
Comment (ceriksson) [reply Sources/PrimerSDK/Classes/User Interface/Root/PrimerNavigationBar.swift:28]: thx, good catch! I'll fix this and do an extra scan for these

**Pull request merged**